pkg/scheduler/partition.go

```diff
@@ -380,7 +406,33 @@ func (pc *PartitionContext) removeApplication(appID string) []*objects.Allocatio
 	// Remove app from queue
 	if queue := app.GetQueue(); queue != nil {
 		queue.RemoveApplication(app)
+
+		var updateGroupMetrics bool
+		if len(app.GetUser().User) > 0 {
+			user := queue.GetUserGroupManager().GetUser(app.GetUser().User)
+			if user != nil {
+				user.DecRunningApplications()
+				g := queue.GetUserGroupManager().GetGroup(user.GetUsedGroup())
+				if g != nil {
+					g.DecRunningApplications()
+					updateGroupMetrics = true
+				}
+			}
+		}
+
+		// Used when limit has been configured only for group, not for individual user
+		if ! updateGroupMetrics && len(app.GetUser().User) > 0 {
+			for _, group := range app.GetUser().Groups {
+				// Is there any group (to which user belongs to) config has limit settings?
+				g := queue.GetUserGroupManager().GetGroup(group)
+				if g != nil {
+					g.DecRunningApplications()
+					break
```

Comment:
This is related to my comment above. If we our pick from multiple groups is deterministic, then it's fine. Otherwise we have to think about how to handle the multiple group case (I don't know how common that is, but `usergroup.UserGroup` is `string[]`, so at least it has theoretical possibility).

Reply:
Ok, this might be a nitpick, but I looked at the existing group cache code.

I think we might have an issue here. Let's say an user "john" who belongs to "qe" group submitted an application which runs for a while. We have limits set for this group. Then "john" is removed from "qe" and the group cache gets updated. The application finishes but `DecRunningApplications()` will not be called.

We need to store this information somewhere else, probably adding an extra field to `Application` and not relying on `app.GetUser().Groups`.

Reply:
cc @wilfred-s @craigcondit

Reply:
> We need to store this information somewhere else, probably adding an extra field to Application and not relying on app.GetUser().Groups.

Please see my earlier comment https://github.com/apache/incubator-yunikorn-core/pull/353#discussion_r780043936.  user.SetUsedGroup() precisely takes care of this issue only.

Reply:
From what I understand, that only works if `updateGroupMetrics = true`. Maybe I'm still not reading it correctly, but if "limit has been configured only for group, not for individual user" (as the comment says), then a modified list of groups will be checked against that saved value. So in my example, "qe" will not be part of if.

Reply:
Please see line no.415.

Reply:
Talked to Wilfred, let's wait for his feedback. You guys will sync up soon.